src/lib/InstantSearch.js

```diff
@@ -296,87 +236,70 @@ See: https://www.algolia.com/doc/guides/building-search-ui/widgets/create-your-o
+    // DerivedHelper scoped into the `index` widgets.
+    const mainHelper = algoliasearchHelper(this.client);
 
-    helper.search = () => {
+    mainHelper.search = () => {
       // This solution allows us to keep the exact same API for the users but
       // under the hood, we have a different implementation. It should be
       // completely transparent for the rest of the codebase. Only this module
       // is impacted.
-      helper.searchOnlyWithDerivedHelpers();
+      return mainHelper.searchOnlyWithDerivedHelpers();
     };
 
     if (this._searchFunction) {
-      this._mainHelperSearch = helper.search.bind(helper);
-      helper.search = () => {
-        const helperSearchFunction = algoliasearchHelper(
+      this._mainHelperSearch = mainHelper.search.bind(mainHelper);
+      mainHelper.search = () => {
+        const mainIndexHelper = this.mainIndex.getHelper();
+        const searchFunctionHelper = algoliasearchHelper(
           {
             search: () => new Promise(noop),
           },
-          helper.state.index,
-          helper.state
+          mainIndexHelper.state.index,
+          mainIndexHelper.state
         );
-        helperSearchFunction.once('search', ({ state }) => {
-          helper.overrideStateWithoutTriggeringChangeEvent(state);
+        searchFunctionHelper.once('search', ({ state }) => {
+          mainIndexHelper.overrideStateWithoutTriggeringChangeEvent(state);
           this._mainHelperSearch();
         });
-        this._searchFunction(helperSearchFunction);
+        this._searchFunction(searchFunctionHelper);
+        return mainHelper;
       };
     }
 
-    this.helper = helper;
-    this.derivedHelper = derivedHelper;
-
-    this._init(helper, helper.state);
-
-    derivedHelper.on('result', ({ results, state }) => {
-      this._render(helper, results, state);
-    });
-
     // Only the "main" Helper emits the `error` event vs the one for `search`
     // and `results` that are also emitted on the derived one.
-    helper.on('error', ({ error }) => {
+    mainHelper.on('error', ({ error }) => {
       this.emit('error', {
         error,
       });
     });
 
-    this._searchStalledTimer = null;
-    this._isSearchStalled = true;
+    this.mainHelper = mainHelper;
 
-    helper.search();
-
-    derivedHelper.on('search', () => {
-      if (!this._isSearchStalled && !this._searchStalledTimer) {
-        this._searchStalledTimer = setTimeout(() => {
-          this._isSearchStalled = true;
-          this._render(
-            helper,
-            derivedHelper.lastResults,
-            derivedHelper.lastResults._state
-          );
-        }, this._stalledSearchDelay);
-      }
+    this.mainIndex.init({
+      instantSearchInstance: this,
+      parent: null,
     });
 
+    mainHelper.search();
+
+    // Keep the previous reference for legacy purpose, some pattern use
+    // the direct Helper access `search.helper` (e.g mutli-index).
+    this.helper = this.mainIndex.getHelper();
```

Comment:
Which kind of warning would you give? "there's no need for this" or "it is wrong when you are using the index widget"?

Reply:
Both are valid. Users probably don't have to access the `helper` on the instance. There are different way to access it, connectors, widgets, events. The second is also valid, because within a multi-index context the `helper` won't be the full picture of the state anymore.